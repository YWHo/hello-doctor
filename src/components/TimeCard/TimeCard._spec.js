import React from 'react';
import dayjs from 'dayjs';
import { shallow } from 'enzyme';
import { TimeCard } from './TimeCard';
import { DAY } from '../../shared/constants';

describe('TimeCard', () => {
  const today = dayjs('2019-06-16T07:59:00');
  const scheduleObj = {
    Id: '40abb954-2f57-4106-61ec-ddf2acfbf8ed',
    Title: 'General Practitioner',
    Name: 'Dr. Mary Aniston',
    AvailableSlots: {
      '9e7bbebf-92b7-4f1b-a6fc-35a72778cd77': '2019-06-16T08:00:00',
      '744022e0-54fb-41b5-b042-b4f8810d7e38': '2019-06-16T12:00:00',
      '87e5128b-d134-4d67-aca2-eef56a800adc': '2019-06-16T17:00:00',
    },
    PictureURL: '/api/Picture/40abb954-2f57-4106-61ec-ddf2acfbf8ed',
  };

  it('render default correctly', () => {
    const wrapper = shallow(
      <TimeCard meetSchedule={scheduleObj} today={today} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('render to start from morning correctly', () => {
    const wrapper = shallow(
      <TimeCard
        meetSchedule={scheduleObj}
        selectedDayPart={DAY.MORNING}
        today={today}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('render to start from afternoon correctly', () => {
    const wrapper = shallow(
      <TimeCard
        meetSchedule={scheduleObj}
        selectedDayPart={DAY.AFTERNOON}
        today={today}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('render to start from evening correctly', () => {
    const wrapper = shallow(
      <TimeCard
        meetSchedule={scheduleObj}
        selectedDayPart={DAY.EVENING}
        today={today}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('render no time available correctly', () => {
    const emptyScheduleObj = {
      ...scheduleObj,
      AvailableSlots: {},
    };
    const wrapper = shallow(
      <TimeCard meetSchedule={emptyScheduleObj} today={today} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('render only those time that are not in the past', () => {
    const currentTime = dayjs('2019-06-16T12:00:00');
    const wrapper = shallow(
      <TimeCard meetSchedule={scheduleObj} today={currentTime} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
