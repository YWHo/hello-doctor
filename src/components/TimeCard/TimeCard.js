import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import dayjs from 'dayjs';
import TimeSlots from '../TimeSlots';
import {
  filterTimeSlotByPartOfDay,
  filterTimePassedNow
} from '../../shared/dateHelper';
import { toFetchProvider, toggleShowingProfile } from '../../state/actions';
import { getSelectedDayPart } from '../../state/selectors';
import {
  DAY_AFTERNOON,
  DAY_EVENING,
  DAY_MORNING
} from '../../shared/constants';

const Container = styled.div`
  background-color: #fff;
  height: 166.95px;
  width: 100%;
  margin-bottom: 5px;
  padding-top: 15px;
  padding-left: 12px;
`;

const Photo = styled.img`
  float: left;
  height: 81px;
  width: 80px;
`;

const NameFrame = styled.div`
  color: #2a95bf;
  font-family: 'Roboto', sans-serif;
  font-size: 22px;
  min-height: 29px;
  line-height: 29px;
  vertical-align: middle;
`;

const TitleFrame = styled.div`
  color: #818181;
  font-size: 14px;
  height: 19px;
  vertical-align: middle;
`;

const TextFrame = styled.div`
  margin-left: 103px;
  height: 81px;
`;

export function TimeCard(props) {
  const {
    meetSchedule,
    selectedDayPart = DAY_MORNING,
    today = dayjs()
  } = props;
  const {
    Id: doctorId,
    Name: name,
    Title: title,
    AvailableSlots: availableSlots = [],
    PictureURL: pictureURL = ''
  } = meetSchedule;
  const fullUrl = `https://frontendchallenge2019.azurewebsites.net/${pictureURL}`;
  const filteredSlotsByPart = filterTimeSlotByPartOfDay(
    availableSlots,
    selectedDayPart
  );
  const filteredSlots = filterTimePassedNow(filteredSlotsByPart, today);

  return (
    <Container>
      <Photo
        src={fullUrl}
        alt="Doctor's photo"
        onClick={() => onOpeningProfile(props, doctorId)}
      />
      <TextFrame>
        <NameFrame onClick={() => onOpeningProfile(props, doctorId)}>
          {name}
        </NameFrame>
        <TitleFrame onClick={() => onOpeningProfile(props, doctorId)}>
          {title}
        </TitleFrame>
      </TextFrame>
      <TimeSlots availableSlots={filteredSlots} />
    </Container>
  );
}

function onOpeningProfile(props, doctorID) {
  const { dToFetchProvider, dToggleShowingProfile } = props;
  dToFetchProvider(doctorID);
  dToggleShowingProfile(true);
}

TimeCard.propTypes = {
  meetSchedule: PropTypes.object.isRequired,
  selectedDayPart: PropTypes.oneOf([DAY_MORNING, DAY_AFTERNOON, DAY_EVENING]),
  today: PropTypes.object
};

const mapStateToProps = state => {
  return {
    selectedDayPart: getSelectedDayPart(state)
  };
};

function mapDispatchToProps(dispatch) {
  return {
    dToFetchProvider: id => dispatch(toFetchProvider(id)),
    dToggleShowingProfile: status => dispatch(toggleShowingProfile(status))
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TimeCard);
