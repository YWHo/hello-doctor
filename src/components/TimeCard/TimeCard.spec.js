import React from 'react';
import { shallow } from 'enzyme';
import { TimeCard } from './TimeCard';
import {
  DAY_MORNING,
  DAY_AFTERNOON,
  DAY_EVENING
} from '../../shared/constants';

describe('RegisterPage', () => {
  const scheduleObj = {
    Id: '40abb954-2f57-4106-61ec-ddf2acfbf8ed',
    Title: 'General Practitioner',
    Name: 'Dr. Mary Aniston',
    AvailableSlots: {
      '9e7bbebf-92b7-4f1b-a6fc-35a72778cd77': '2019-06-16T08:00:00',
      '744022e0-54fb-41b5-b042-b4f8810d7e38': '2019-06-16T12:00:00',
      '87e5128b-d134-4d67-aca2-eef56a800adc': '2019-06-16T17:00:00'
    },
    PictureURL: '/api/Picture/40abb954-2f57-4106-61ec-ddf2acfbf8ed'
  };

  it('render default correctly', () => {
    const wrapper = shallow(<TimeCard meetSchedule={scheduleObj} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('render to start from morning correctly', () => {
    const wrapper = shallow(
      <TimeCard meetSchedule={scheduleObj} selectedDayPart={DAY_MORNING} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('render to start from afternoon correctly', () => {
    const wrapper = shallow(
      <TimeCard meetSchedule={scheduleObj} selectedDayPart={DAY_AFTERNOON} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('render to start from evening correctly', () => {
    const wrapper = shallow(
      <TimeCard meetSchedule={scheduleObj} selectedDayPart={DAY_EVENING} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('render no time available correctly', () => {
    const emptyScheduleObj = {
      ...scheduleObj,
      AvailableSlots: {}
    };
    const wrapper = shallow(<TimeCard meetSchedule={emptyScheduleObj} />);
    expect(wrapper).toMatchSnapshot();
  });
});
