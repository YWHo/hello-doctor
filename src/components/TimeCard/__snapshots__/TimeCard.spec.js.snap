// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeCard render default correctly 1`] = `
<styled.div>
  <styled.img
    alt="Doctor's photo"
    onClick={[Function]}
    src="https://frontendchallenge2019.azurewebsites.net//api/Picture/40abb954-2f57-4106-61ec-ddf2acfbf8ed"
  />
  <styled.div>
    <styled.div
      onClick={[Function]}
    >
      Dr. Mary Aniston
    </styled.div>
    <styled.div
      onClick={[Function]}
    >
      General Practitioner
    </styled.div>
  </styled.div>
  <Connect(TimeSlots)
    availableSlots={
      Object {
        "744022e0-54fb-41b5-b042-b4f8810d7e38": "2019-06-16T12:00:00",
        "87e5128b-d134-4d67-aca2-eef56a800adc": "2019-06-16T17:00:00",
        "9e7bbebf-92b7-4f1b-a6fc-35a72778cd77": "2019-06-16T08:00:00",
      }
    }
  />
</styled.div>
`;

exports[`TimeCard render no time available correctly 1`] = `
<styled.div>
  <styled.img
    alt="Doctor's photo"
    onClick={[Function]}
    src="https://frontendchallenge2019.azurewebsites.net//api/Picture/40abb954-2f57-4106-61ec-ddf2acfbf8ed"
  />
  <styled.div>
    <styled.div
      onClick={[Function]}
    >
      Dr. Mary Aniston
    </styled.div>
    <styled.div
      onClick={[Function]}
    >
      General Practitioner
    </styled.div>
  </styled.div>
  <Connect(TimeSlots)
    availableSlots={Object {}}
  />
</styled.div>
`;

exports[`TimeCard render only those time that are not in the past 1`] = `
<styled.div>
  <styled.img
    alt="Doctor's photo"
    onClick={[Function]}
    src="https://frontendchallenge2019.azurewebsites.net//api/Picture/40abb954-2f57-4106-61ec-ddf2acfbf8ed"
  />
  <styled.div>
    <styled.div
      onClick={[Function]}
    >
      Dr. Mary Aniston
    </styled.div>
    <styled.div
      onClick={[Function]}
    >
      General Practitioner
    </styled.div>
  </styled.div>
  <Connect(TimeSlots)
    availableSlots={
      Object {
        "87e5128b-d134-4d67-aca2-eef56a800adc": "2019-06-16T17:00:00",
      }
    }
  />
</styled.div>
`;

exports[`TimeCard render to start from afternoon correctly 1`] = `
<styled.div>
  <styled.img
    alt="Doctor's photo"
    onClick={[Function]}
    src="https://frontendchallenge2019.azurewebsites.net//api/Picture/40abb954-2f57-4106-61ec-ddf2acfbf8ed"
  />
  <styled.div>
    <styled.div
      onClick={[Function]}
    >
      Dr. Mary Aniston
    </styled.div>
    <styled.div
      onClick={[Function]}
    >
      General Practitioner
    </styled.div>
  </styled.div>
  <Connect(TimeSlots)
    availableSlots={
      Object {
        "744022e0-54fb-41b5-b042-b4f8810d7e38": "2019-06-16T12:00:00",
        "87e5128b-d134-4d67-aca2-eef56a800adc": "2019-06-16T17:00:00",
      }
    }
  />
</styled.div>
`;

exports[`TimeCard render to start from evening correctly 1`] = `
<styled.div>
  <styled.img
    alt="Doctor's photo"
    onClick={[Function]}
    src="https://frontendchallenge2019.azurewebsites.net//api/Picture/40abb954-2f57-4106-61ec-ddf2acfbf8ed"
  />
  <styled.div>
    <styled.div
      onClick={[Function]}
    >
      Dr. Mary Aniston
    </styled.div>
    <styled.div
      onClick={[Function]}
    >
      General Practitioner
    </styled.div>
  </styled.div>
  <Connect(TimeSlots)
    availableSlots={
      Object {
        "87e5128b-d134-4d67-aca2-eef56a800adc": "2019-06-16T17:00:00",
      }
    }
  />
</styled.div>
`;

exports[`TimeCard render to start from morning correctly 1`] = `
<styled.div>
  <styled.img
    alt="Doctor's photo"
    onClick={[Function]}
    src="https://frontendchallenge2019.azurewebsites.net//api/Picture/40abb954-2f57-4106-61ec-ddf2acfbf8ed"
  />
  <styled.div>
    <styled.div
      onClick={[Function]}
    >
      Dr. Mary Aniston
    </styled.div>
    <styled.div
      onClick={[Function]}
    >
      General Practitioner
    </styled.div>
  </styled.div>
  <Connect(TimeSlots)
    availableSlots={
      Object {
        "744022e0-54fb-41b5-b042-b4f8810d7e38": "2019-06-16T12:00:00",
        "87e5128b-d134-4d67-aca2-eef56a800adc": "2019-06-16T17:00:00",
        "9e7bbebf-92b7-4f1b-a6fc-35a72778cd77": "2019-06-16T08:00:00",
      }
    }
  />
</styled.div>
`;
